File: backend/api/app/services/ai_service.py
"""
============================================================================
FORGE — Framework for Observed Rules, Generation & Engineered Data
============================================================================

File: ai_service.py
Purpose: Orchestrates FORGE AI provider capabilities.

Author: Ranjoy Sen
Email: ranjoy.sen@collins.com

============================================================================
"""

from app.interfaces.ai_provider import (
    AIDataModelProposal,
    AIFieldProposal,
    AISemanticPreview,
    AIProvider,
    AIProviderStatus,
)


class AIService:
    """Provides application-level AI capabilities."""

    def __init__(self, provider: AIProvider) -> None:
        self._provider = provider

    def get_capabilities(self) -> AIProviderStatus:
        """Return the current AI capability status."""

        return self._provider.get_status()

    def suggest_data_model(
        self,
        prompt: str,
    ) -> AIDataModelProposal:
        """Generate a Data Model proposal from user intent."""

        return self._provider.suggest_data_model(prompt)


    def preview_semantic_values(
        self,
        description: str,
    ) -> AISemanticPreview:
        """Generate representative semantic field values for preview."""

        return self._provider.preview_semantic_values(description)

    def propose_field(
        self,
        mode: str,
        entity_name: str,
        request: str,
        existing_field: dict[str, object] | None = None,
    ) -> AIFieldProposal:
        """Generate a structured FORGE field proposal."""

        return self._provider.propose_field(
            mode=mode,
            entity_name=entity_name,
            request=request,
            existing_field=existing_field,
        )

File: frontend/src/app/core/services/ai.service.ts
/**
 * File: ai.service.ts
 * Purpose: Provides frontend access to FORGE AI capabilities.
 *
 * Author: Ranjoy Sen
 * Email: ranjoy.sen@collins.com
 */

import {
  HttpClient,
  HttpContext,
} from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable, map } from 'rxjs';

import { AICapability } from '../interfaces/ai-capability.interface';
import { AISemanticPreview } from '../interfaces/ai-semantic-preview.interface';
import { AIDataModelProposal } from '../interfaces/ai-data-model-proposal.interface';
import { AIDataModelProposalResponse } from '../interfaces/ai-data-model-proposal-response.interface';
import { ApiLoadingMessage } from '../enums/api-loading-message.enum';
import { API_LOADING_MESSAGE } from '../tokens/api-loading-message.token';
import { API_LOADING_SKIP } from '../tokens/api-loading-skip.token';

@Injectable({
  providedIn: 'root',
})
export class AIService {
  private readonly http = inject(HttpClient);

  getCapabilities(): Observable<AICapability> {
    return this.http.get<AICapability>(
      '/api/v1/ai/capabilities',
      {
        context: new HttpContext().set(
          API_LOADING_SKIP,
          true,
        ),
      },
    );
  }

  previewSemanticValues(
    description: string,
  ): Observable<AISemanticPreview> {
    return this.http.post<{
      status: AISemanticPreview['status'];
      message: string;
      preview_values: string[];
    }>(
      '/api/v1/ai/semantic/preview',
      { description },
      {
        context: new HttpContext().set(
          API_LOADING_MESSAGE,
          ApiLoadingMessage.GeneratingSemanticPreviewWithAI,
        ),
      },
    ).pipe(
      map((response) => ({
        status: response.status,
        message: response.message,
        previewValues: response.preview_values,
      })),
    );
  }

  suggestDataModel(prompt: string): Observable<AIDataModelProposal> {
    return this.http
      .post<AIDataModelProposalResponse>(
        '/api/v1/ai/data-models/suggest',
        { prompt },
        {
          context: new HttpContext().set(
            API_LOADING_MESSAGE,
            ApiLoadingMessage.GeneratingDataModelWithAI,
          ),
        },
      )
      .pipe(
        map((response) => ({
          name: response.name,
          description: response.description,
          suggestedTags: response.suggested_tags,
          reasoning: response.reasoning,
        })),
      );
  }
}
